diff --git a/internal/lib/container_server.go b/internal/lib/container_server.go
index 673d249d3..5052747c4 100644
--- a/internal/lib/container_server.go
+++ b/internal/lib/container_server.go
@@ -412,7 +412,7 @@ func (c *ContainerServer) LoadContainer(ctx context.Context, id string) (retErr
 
 	var someNameOfTheImage *bundle.BundleName
 	if s, ok := m.Annotations[annotations.SomeNameOfTheImage]; ok && s != "" {
-		name, err := bundle.ParseBundleName(s)
+		name, err := bundle.PaserDockerName(s)
 		if err != nil {
 			return fmt.Errorf("invalid %s annotation %q: %w", annotations.SomeNameOfTheImage, s, err)
 		}
diff --git a/internal/storage/runtime.go b/internal/storage/runtime.go
index 5c799d402..0630a041f 100644
--- a/internal/storage/runtime.go
+++ b/internal/storage/runtime.go
@@ -6,6 +6,7 @@ import (
 	"fmt"
 	"os"
 	"path/filepath"
+	"strings"
 	"time"
 
 	json "github.com/json-iterator/go"
@@ -134,7 +135,14 @@ func (ss *StorageService) createContainerOrPodSandbox(containerID string, templa
 	now := time.Now()
 	metadata.CreatedAt = now.Unix()
 
-	id, rootFs, imgConfig, err := ss.bm.CreateContainerById(template.imageID)
+	bundleId := template.imageID
+	{
+		id := strings.SplitN(bundleId.String(), ":", 2)
+		if len(id) == 2 {
+			bundleId = bundle.BundleId(id[1])
+		}
+	}
+	id, rootFs, imgConfig, err := ss.bm.CreateContainerById(bundleId)
 	if err != nil {
 		if metadata.Pod {
 			logrus.Debugf("Failed to create pod sandbox %s(%s): %v", metadata.PodName, metadata.PodID, err)
@@ -193,7 +201,7 @@ func (ss *StorageService) createContainerOrPodSandbox(containerID string, templa
 		return ContainerInfo{}, err
 	}
 
-	return ContainerInfo{
+	info := ContainerInfo{
 		ID:           id,
 		Names:        []string{},
 		ImageID:      template.imageID.String(),
@@ -204,7 +212,19 @@ func (ss *StorageService) createContainerOrPodSandbox(containerID string, templa
 		Metadata:     string(mdata),
 		ProcessLabel: "",
 		MountLabel:   "",
-	}, nil
+	}
+
+	err = ss.saveInfo(containerID, info)
+	if err != nil {
+		return ContainerInfo{}, err
+	}
+
+	err = ss.saveInfo(id, info)
+	if err != nil {
+		return ContainerInfo{}, err
+	}
+
+	return info, nil
 }
 
 // DeleteContainer deletes a container, unmounting it first if need be.
@@ -216,16 +236,25 @@ func (ss *StorageService) DeleteContainer(ctx context.Context, idOrName string)
 	if idOrName == "" {
 		return ErrInvalidContainerID
 	}
-	err := ss.bm.DeleteContainer(idOrName)
+
+	info, err := ss.loadInfo(idOrName)
+	if err != nil {
+		log.Warnf(ctx, "Failed to get container info %s: %v", idOrName, err)
+		return err
+	}
+
+	err = ss.bm.DeleteContainer(info.ID)
 	if err != nil {
 		log.Debugf(ctx, "Failed to delete container %q: %v", idOrName, err)
 		return err
 	}
+
 	infoFile := filepath.Join(ss.info, idOrName)
 	err = os.Remove(infoFile)
 	if err != nil && !os.IsNotExist(err) {
 		return fmt.Errorf("failed to delete metadata file: %w", err)
 	}
+
 	return nil
 }
 
diff --git a/internal/storage/storage.go b/internal/storage/storage.go
index 643bd33bc..d6cc1999d 100644
--- a/internal/storage/storage.go
+++ b/internal/storage/storage.go
@@ -6,7 +6,6 @@ import (
 	"os"
 	"path/filepath"
 	"regexp"
-	"strconv"
 	"strings"
 
 	"github.com/L-F-Z/TaskC/pkg/bundle"
@@ -60,7 +59,7 @@ func (ss *StorageService) ListImages() (result []*types.Image, err error) {
 		uid, username := getUser(bundle.Blueprint.User)
 		img := &types.Image{
 			Id:          fmt.Sprintf("sha256:%s", bundle.Id),
-			RepoTags:    []string{},
+			RepoTags:    []string{fmt.Sprintf("%s:%s", bundle.Blueprint.Name, bundle.Blueprint.Version)},
 			RepoDigests: []string{fmt.Sprintf("%s@sha256:%s", bundle.Blueprint.Name, bundle.Id)},
 			Size_:       bundle.Size,
 			Uid:         &types.Int64Value{Value: *uid},
@@ -75,20 +74,23 @@ func (ss *StorageService) ListImages() (result []*types.Image, err error) {
 // getUserFromImage gets uid or user name of the image user.
 // If user is numeric, it will be treated as uid; or else, it is treated as user name.
 func getUser(user string) (id *int64, username string) {
-	// return both empty if user is not specified in the image.
-	if user == "" {
-		return nil, ""
-	}
-	// split instances where the id may contain user:group
-	user = strings.Split(user, ":")[0]
-	// user could be either uid or user name. Try to interpret as numeric uid.
-	uid, err := strconv.ParseInt(user, 10, 64)
-	if err != nil {
-		// If user is non numeric, assume it's user name.
-		return nil, user
-	}
-	// If user is a numeric uid.
-	return &uid, ""
+	// // return both empty if user is not specified in the image.
+	// if user == "" {
+	// 	user = "0:0"
+	// }
+	// // split instances where the id may contain user:group
+	// user = strings.Split(user, ":")[0]
+	// // user could be either uid or user name. Try to interpret as numeric uid.
+	// uid, err := strconv.ParseInt(user, 10, 64)
+	// if err != nil {
+	// 	// If user is non numeric, assume it's user name.
+	// 	uid = 0
+	// 	return &uid, user
+	// }
+	// // If user is a numeric uid.
+	// return &uid, ""
+	var uid int64 = 0
+	return &uid, "root"
 }
 
 // ImageStatusByID returns status of a single image
@@ -100,9 +102,9 @@ func (ss *StorageService) ImageStatusByID(id bundle.BundleId) (img *types.Image,
 	uid, username := getUser(bundle.Blueprint.User)
 	img = &types.Image{
 		Id:          fmt.Sprintf("sha256:%s", bundle.Id),
-		RepoTags:    []string{},
+		RepoTags:    []string{fmt.Sprintf("%s:%s", bundle.Blueprint.Name, bundle.Blueprint.Version)},
 		RepoDigests: []string{fmt.Sprintf("%s@sha256:%s", bundle.Blueprint.Name, bundle.Id)},
-		Size_:       0,
+		Size_:       bundle.Size,
 		Uid:         &types.Int64Value{Value: *uid},
 		Username:    username,
 		Pinned:      false,
@@ -110,6 +112,14 @@ func (ss *StorageService) ImageStatusByID(id bundle.BundleId) (img *types.Image,
 	return
 }
 
+func (ss *StorageService) GetBundleByID(id bundle.BundleId) (bundle *bundle.Bundle, err error) {
+	return ss.bm.GetById(id)
+}
+
+func (ss *StorageService) GetBundleByName(name bundle.BundleName) (bundle *bundle.Bundle, err error) {
+	return ss.bm.Get(name.Name, name.Version)
+}
+
 // ImageStatusByName returns status of an image tagged with name.
 func (ss *StorageService) ImageStatusByName(name bundle.BundleName) (img *types.Image, err error) {
 	bundle, err := ss.bm.Get(name.Name, name.Version)
@@ -119,9 +129,9 @@ func (ss *StorageService) ImageStatusByName(name bundle.BundleName) (img *types.
 	uid, username := getUser(bundle.Blueprint.User)
 	img = &types.Image{
 		Id:          fmt.Sprintf("sha256:%s", bundle.Id),
-		RepoTags:    []string{},
+		RepoTags:    []string{fmt.Sprintf("%s:%s", bundle.Blueprint.Name, bundle.Blueprint.Version)},
 		RepoDigests: []string{fmt.Sprintf("%s@sha256:%s", bundle.Blueprint.Name, bundle.Id)},
-		Size_:       0,
+		Size_:       bundle.Size,
 		Uid:         &types.Int64Value{Value: *uid},
 		Username:    username,
 		Pinned:      false,
diff --git a/pkg/config/config.go b/pkg/config/config.go
index 3b2142227..3820fb3c3 100644
--- a/pkg/config/config.go
+++ b/pkg/config/config.go
@@ -1377,7 +1377,7 @@ func validateExecutablePath(executable, currentPath string) (string, error) {
 
 // ParsePauseImage parses the .PauseImage value as into a validated, well-typed value.
 func (c *ImageConfig) ParsePauseImage() bundle.BundleName {
-	name, _ := bundle.ParseBundleName(c.PauseImage)
+	name, _ := bundle.PaserDockerName(c.PauseImage)
 	return name
 }
 
diff --git a/server/container_create_linux.go b/server/container_create_linux.go
index 952d3f2f8..5ea2bd6e5 100644
--- a/server/container_create_linux.go
+++ b/server/container_create_linux.go
@@ -85,14 +85,37 @@ func (s *Server) createSandboxContainer(ctx context.Context, ctr ctrfactory.Cont
 		return nil, err
 	}
 
-	bundleName, err := bundle.ParseBundleName(userRequestedImage)
-	if err != nil {
-		return nil, err
-	}
+	var bundleName bundle.BundleName
+	var imgResult *types.Image
 
-	imgResult, err := s.StorageService().ImageStatusByName(bundleName)
-	if err != nil {
-		return nil, err
+	if strings.HasPrefix(userRequestedImage, "sha256:") {
+		id := bundle.BundleId(strings.TrimPrefix(userRequestedImage, "sha256:"))
+
+		imgResult, err = s.StorageService().ImageStatusByID(id)
+		if err != nil {
+			return nil, err
+		}
+
+		b, err := s.StorageService().GetBundleByID(id)
+		if err != nil {
+			return nil, err
+		}
+
+		bundleName = bundle.BundleName{
+			Name:    b.Blueprint.Name,
+			Version: b.Blueprint.Version,
+		}
+
+	} else {
+		bundleName, err = bundle.PaserDockerName(userRequestedImage)
+		if err != nil {
+			return nil, err
+		}
+
+		imgResult, err = s.StorageService().ImageStatusByName(bundleName)
+		if err != nil {
+			return nil, err
+		}
 	}
 
 	imageID := bundle.BundleId(imgResult.Id)
@@ -373,7 +396,7 @@ func (s *Server) createSandboxContainer(ctx context.Context, ctr ctrfactory.Cont
 	created := time.Now()
 	seccompRef := types.SecurityProfile_Unconfined.String()
 
-	if err := s.FilterDisallowedAnnotations(sb.Annotations(), imgResult.Spec.Annotations, sb.RuntimeHandler()); err != nil {
+	if err := s.FilterDisallowedAnnotations(sb.Annotations(), imgResult.GetSpec().GetAnnotations(), sb.RuntimeHandler()); err != nil {
 		return nil, fmt.Errorf("filter image annotations: %w", err)
 	}
 
@@ -388,7 +411,7 @@ func (s *Server) createSandboxContainer(ctx context.Context, ctr ctrfactory.Cont
 			containerID,
 			ctr.Config().Metadata.Name,
 			sb.Annotations(),
-			imgResult.Spec.Annotations,
+			imgResult.GetSpec().GetAnnotations(),
 			specgen,
 			securityContext.Seccomp,
 		)
diff --git a/server/image_pull.go b/server/image_pull.go
index 1a4582697..4ccf5c5d2 100644
--- a/server/image_pull.go
+++ b/server/image_pull.go
@@ -93,7 +93,7 @@ func (s *Server) pullImage(ctx context.Context, pullArgs *pullArguments) (bundle
 	ctx, span := log.StartSpan(ctx)
 	defer span.End()
 
-	name, err := bundle.ParseBundleName(pullArgs.image)
+	name, err := bundle.PaserDockerName(pullArgs.image)
 	if err != nil {
 		return "", err
 	}
diff --git a/server/image_remove.go b/server/image_remove.go
index b4a605efd..5694625b4 100644
--- a/server/image_remove.go
+++ b/server/image_remove.go
@@ -6,7 +6,6 @@ import (
 
 	types "k8s.io/cri-api/pkg/apis/runtime/v1"
 
-	"github.com/L-F-Z/TaskC/pkg/bundle"
 	"github.com/L-F-Z/cri-t/internal/log"
 )
 
@@ -32,10 +31,16 @@ func (s *Server) removeImage(ctx context.Context, imageRef string) (untagErr err
 	_, span := log.StartSpan(ctx)
 	defer span.End()
 
-	name, err := bundle.ParseBundleName(imageRef)
-	if err != nil {
-		return err
-	}
-	// TODO: Add --image-volume support
-	return s.StorageService().UntagImage(name)
+	// XXX: imageRef will be a prefix of IMAGE-ID
+	// Raw ImageID: sha256:aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee
+	// Possible input: aaaaaaaa-bb, aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee
+
+	return errors.New("not implemented")
+
+	// name, err := bundle.ParseBundleName(imageRef)
+	// if err != nil {
+	// 	return err
+	// }
+	// // TODO: Add --image-volume support
+	// return s.StorageService().UntagImage(name)
 }
diff --git a/server/image_status.go b/server/image_status.go
index d2afd73d3..e80f84b1e 100644
--- a/server/image_status.go
+++ b/server/image_status.go
@@ -38,7 +38,7 @@ func (s *Server) ImageStatus(ctx context.Context, req *types.ImageStatusRequest)
 			RepoDigests: status.RepoDigests,
 			Size_:       uint64(status.Size()),
 			Spec: &types.ImageSpec{
-				Annotations: status.Spec.Annotations,
+				Annotations: status.GetSpec().GetAnnotations(),
 			},
 			Pinned: status.Pinned,
 		},
@@ -59,11 +59,11 @@ func (s *Server) ImageStatus(ctx context.Context, req *types.ImageStatusRequest)
 // storageImageStatus calls ImageStatus for a k8s ImageSpec.
 // Returns (nil, nil) if image was not found.
 func (s *Server) storageImageStatus(ctx context.Context, spec types.ImageSpec) (*types.Image, error) {
-	bundleName, err := bundle.ParseBundleName(spec.Image)
+	bundleName, err := bundle.PaserDockerName(spec.Image)
 	if err == nil {
 		return s.StorageService().ImageStatusByName(bundleName)
 	}
-	bundleName, err = bundle.ParseBundleName(spec.UserSpecifiedImage)
+	bundleName, err = bundle.PaserDockerName(spec.UserSpecifiedImage)
 	if err == nil {
 		return s.StorageService().ImageStatusByName(bundleName)
 	}
diff --git a/server/sandbox_run_linux.go b/server/sandbox_run_linux.go
index 9cd67c649..cf4873e01 100644
--- a/server/sandbox_run_linux.go
+++ b/server/sandbox_run_linux.go
@@ -317,10 +317,6 @@ func (s *Server) runPodSandbox(ctx context.Context, req *types.RunPodSandboxRequ
 	// bind mount the pod shm
 	g.AddMount(mnt)
 
-	err = s.setPodSandboxMountLabel(ctx, sboxID, mountLabel)
-	if err != nil {
-		return nil, err
-	}
 
 	if err := s.CtrIDIndex().Add(sboxID); err != nil {
 		return nil, err
diff --git a/vendor/github.com/L-F-Z/TaskC/pkg/bundle/manager.go b/vendor/github.com/L-F-Z/TaskC/pkg/bundle/manager.go
index ab67b4790..90d58ebb6 100644
--- a/vendor/github.com/L-F-Z/TaskC/pkg/bundle/manager.go
+++ b/vendor/github.com/L-F-Z/TaskC/pkg/bundle/manager.go
@@ -50,6 +50,23 @@ func ParseBundleName(str string) (BundleName, error) {
 	}
 }
 
+func PaserDockerName(uri string) (BundleName, error) {
+	splitter := func(uri string) (full, version string) {
+		idx := strings.LastIndex(uri, ":")
+		if idx == -1 {
+			full = uri
+			version = "latest"
+		} else {
+			full = uri[:idx]
+			version = uri[idx+1:]
+		}
+		return
+	}
+
+	full, version := splitter(uri)
+	return BundleName{Name: full, Version: version}, nil
+}
+
 type BundleId string
 
 func (b BundleId) String() string {
